### DISCUSSION TITLE
It is working right?

### DISCUSSION THREAD

**Discussion by gabrielustosa** [fastapi/fastapi#11303] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class QueryParams(BaseModel):
    test: str

    @field_validator('test')
    def check_test_starts_with(cls, v):
        if not v.startswith("test"):
            raise ValueError('O campo "test" deve começar com "test"')
        return v

@app.get('/test')
def test(params: QueryParams = Depends()):
   ...
```

### Description

I have an endpoint that uses a Pydantic model for query parameters. This model includes validation for one of its fields. However, instead of returning a 422 invalid request error (as it would when using a Pydantic model to validate the request body), it returns a 500 error. Is it supposed to work like this?

### Operating System

Windows

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

Python 3.11.0rc1

**Comment by khaledadrani** [comment]
You can work around this by creating a custom exception and adding its custom handler to your fastapi application. 
Check this [stackoverflow question](https://stackoverflow.com/questions/72560837/custom-fastapi-query-parameter-validation) for more information.

````
from fastapi import FastAPI, Depends
from pydantic.v1 import BaseModel, validator
from fastapi.requests import Request
from fastapi.responses import JSONResponse


class BadQueryParam(Exception):
    pass


class QueryParams(BaseModel):
    test: str

    @validator('test')
    def check_test_starts_with(cls, v):
        if not v.startswith("test"):
            raise BadQueryParam('O campo "test" deve começar com "test"')
        return v


app = FastAPI()


@app.exception_handler(BadQueryParam)
async def value_error_exception_handler(request: Request, exc: BadQueryParam):
    return JSONResponse(
        status_code=422,
        content={"message": str(exc)},
    )


@app.get('/test')
def test(params: QueryParams = Depends()):
    return {"test": params.test}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run("app:app", host="localhost", port=8000)

````

**Comment by YuriiMotov** [comment]
I believe it shouldn't return 500 error, but now it really does.

But if you use Field's `pattern` parameter instead of `field_validator`(which is IMO better in this particular case), it works fine:
```
from fastapi import Depends, FastAPI
from pydantic import BaseModel, Field

app = FastAPI()

class QueryParams(BaseModel):
    test: str = Field(pattern=r'^test')

@app.get('/test')
def test(params: QueryParams = Depends()):
   ...
```

**Comment by YuriiMotov** [comment, marked as answer]
Reviewed this one more time.

Since v0.115.0 FastAPI officially supports Pydantic models in `Query` parameters (and in `Header` and `Cookie` as well).

So, after updating the code according to new docs, it works well:

```python
from typing import Annotated
from fastapi import FastAPI, Query
from pydantic import BaseModel, field_validator


class QueryParams(BaseModel):
    test: str

    @field_validator("test")
    def check_test_starts_with(cls, v):
        if not v.startswith("test"):
            raise ValueError('O campo "test" deve começar com "test"')
        return v


app = FastAPI()


@app.get("/test")
def test(params: Annotated[QueryParams, Query()]):
    pass

```
